README.md
# jjDicTrans
Simple firefox extenstion for English to Korean dictionary &amp; translator

Daum/Naver/Google등의 Backend를 활용한 간단한 영한사전 + 영한번역 Addon 입니다.
영어공부를 할때 불편해서 만들어 봤습니다. 많이 활용하시고 피드백 부탁드립니다.

## 활성화키 설정
<kbd>Shift</kbd> <kbd>Ctrl</kbd> <kbd>Alt</kbd> 와 Function Key와 같은 기타 키를 조합해서 원하는 단축키를 만들어서 사용하면 된다. 또한, 마우스 Double Click 조합이 가능하다 (Double Click 단독으로도 사용가능). 최초 기본값은 <kbd>Shift</kbd> 이다.

## 사전기능 사용하기
검색을 원하는 단어 위에 마우스를 위치시키고 "활성화키"를 누르거나 Drag해서 특정 단어를 선택한 상태에서  "활성화키"를 누르면 된다.

## 번역기능 사용하기
번역하고자 하는 문장을 Drag해서 선택한 후 "활성화키"를 누르거나 선택 후 마우스 오른쪽 버튼을 눌러서 나타나는 컨택스트 메뉴를 클릭하면 된다. 단, 만약 Drag로 선택된 문장이 하나의 단어라면 사전 기능이 활성화 된다.

## 설정변경
버튼 영역의 jjDicTrans 버튼을 클릭하면 설정창이 나타나며 변경된 설정은 바로 적용된다.

* **사용여부**

	> 이 설정이 Check 되어 있을때만 jjDicTrans의 기능이 동작하게 된다.

* **실행모드**

	> 특정기능만 사용하고 싶을때 설정 할 수 있도록 해주는 옵션이다.

	```
	- 사전 + 번역
	- 사전 only
	- 번역 only
	```

* **활설화 키 조합**

	> 사전을 활성화 시켜줄 Key를 설정한다.

	> 만약 마우스 더블클릭이 선택되지 않았다면 4개의 키 중에서 최소한 하나는 선택되어야 한다.

* **마우스 더블클릭**

	> 활성화 키와 함께 마우스 더블클릭을 조할 할 수 있다. 활성화 키 없이 마우스 더블 클릭 만으로도 설정 가능 하다.

* **사전 Backend**

	```
	- Daum 미니 사전
	- Naver 미니 사전
	```
* **번역 Backend**

	```
	- Google 번역
	- Naver 번역
	```

* **Popup 위치**

	> 사전/번역 정보 Popup이 출력 되는 위치에 대한 설정이다. 팜업창이 컨텐츠를 가려서 불편하다면 설정을 변경해 보자.

	```
	- 우측상단
	- 커서위치
	```


## ChangeLog
```
v0.6.0
	- Add Quick search function

v0.5.4
	- 사전창에 타이틀이 다시 보이도록 수정
	- bugfix
		- mode 와 popup_pos 변경 값이 설정창에 반영되지 않던 문제 수정

v0.5.3
	- backgrun-color --> background-color

v0.5.2
	- adjust popup size

v0.5.1
	- Popup창 위치 변경 후 원복 되지 않던 문제 수정

v0.5.0
	- Popup창 위치를 설정할 수 있는 옵션 추가.
	- Mode 설정 옵션 추가
		- 사전 + 번역 모드
		- 사전 Only 모드
		- 번역 Only 모드
v0.4.2
	- 번역과사전 창을 임의로 띄울수 있는 링크 추가
v0.4.1
  - Popup Context Menu에 여러 항목이 중복으로 나열되는 문제 처리
v0.4.0
  - 선택영역에서 Mouse 오른쪽 버튼의 Context Menu에서 번역하는 Menu 항목 추가.
  - Popup Panel 위치 수정
  - Popup 내용에 집중 할 수 있도록 몇몇 항목 제거
v0.3.0
  - 키 충돌로 인해 불필요한 창이 뜨는 경우가 있다는 의견이 있어 단축키를 설정가능하도록 변경 했습니다.
  - 설정을 Addon 페이지가 아닌 버튼 영역에서 할 수있도록 변경 했습니다.
```

## Links
  * https://addons.mozilla.org/ko/firefox/addon/jjdictrans-simple-%EC%98%81%ED%95%9C%EC%82%AC%EC%A0%84%EB%B2%88%EC%97%AD/?src=ss


ChangeLog
Changes to 0.6.0 (2016-07-18)
	- Add Quick search function

Changes to 0.5.4 (2016-03-31)
	- 사전창에 타이틀이 다시 보이도록 수정
	- build tool changed (from cfx to jpm)
	- bugfix
		- mode 와 popup_pos 변경 값이 설정창에 반영되지 않던 문제 수정

Changes to 0.5.3 (2016-01-08)
	- bugfix
		- backgrun-color --> background-color

Changes to 0.5.2 (2016-01-08)
	- adjust popup size

Changes to 0.5.1 (2015-11-30)
	- Popup창 위치 변경 후 원복 되지 않던 문제 수정

Changes to 0.5.0 (2015-11-26)
	- Popup창 위치를 설정할 수 있는 옵션 추가.
	- 실행Mode 설정 옵션 추가
		- 사전 + 번역 모드
		- 사전 Only 모드
		- 번역 Only 모드

Changes to 0.4.2 (2015-10-18)
	- 번역과사전 창을 임의로 띄울수 있는 링크 추가

Changes to 0.4.1 (2015-09-08)
	- Popup Context Menu에 여러 항목이 중복으로 나열되는 문제 처리

Changes to 0.4.0 (2015-09-02)
	- 선택영역에서 Mouse 오른쪽 버튼의 Context Menu에서 번역하는 Menu 항목 추가.
	- Popup Panel 위치 수정
	- Popup 내용에 집중 할 수 있도록 몇몇 항목 제거

Changes to 0.3.0 (2015-08-25)
	- 설정을 Panel쪽으로 이동.
	- 활성화키 다양화
	- 마우스 Double Click 기능 추가.

Changes to 0.2.0 (2015-05-21)
	- add On/Off action button
	- translation 안정성 향상

Changes to 0.1.3 (2015-05-16)
	- 0.1.2 버전에서 선택영역에 변형을 주는 문제점이 있어서 수정.

Changes to 0.1.2 (2015-05-16)
	- 특정 문장이 번역으로 읽혀지지 않는 문제 해결 (공백반복/\r/\n을 공백문자로 변환)

Changes to 0.1.1 (2015-05-11)
	- 활성화Key(Ctrl,Shift)+MouseClick 시에 Panel이 뜨지 않도록 수정
	- 선택된 단어가 영어가 아닌경우에 창을 보여주지 않도록 수정(사전에 경우에만 해당, 번역은 그대로 보여준다.)
	- 활성화Key 기본값을 Ctrl 에서 Shift로 변경
	- Panel 의 위치 수정

Changes to 0.1 (2015-05-01)
	- 0.1 Release


lib/main.js
/**
 * jjDict namespace.
 */

var backendSet = {
	en:{
		dict:{
			naver:{
				url:function(txt){
					var a = new Array();
					a.push("http://endic.naver.com/popManager.nhn?sLn=kr&m=search&query="+encodeURI(txt)+"&searchOption=entry_idiom&isOnlyViewEE=N");
					a.push(false);
					return a;
				},
      			size:{w:410, h:440 }
			},
			daum:{
				url:function(txt){
					var a = new Array();
					a.push("http://small.dic.daum.net/search.do?q="+encodeURI(txt)+"&dic=eng&search_first=Y");
					a.push(false);
					return a;
				},
				size:{w:410, h:440 }
			}
		},
		trans:{
			naver:{
				url:function(txt){
					var id="transEditorText";
					var a = new Array();
					a.push("http://translate.naver.com/#/en/ko/");
					a.push({oid:id,text:txt,btn:"startTranslateBtn"});
					return a;
				},
      			size:{w:900, h:420 }
			},
			google:{
				url:function(txt){
					var id="source";
					var a = new Array();
					a.push("https://translate.google.co.kr/#en/ko/");
					a.push({oid:id,text:txt});
					return a;
				},
      			size:{w:800, h:420 }
			}
		}
	}
};
var workers = new Array();
function detachWorker(worker, workerArray) {
	var index = workerArray.indexOf(worker);
	if (index != -1) {
		workerArray.splice(index,1);
	}
}
function iconStateChange(button, state) {
	if (state == true) {
		button.state("window",{
			"icon" : {
				"16":require("sdk/self").data.url("icons/icon_16.png"),
				"32":require("sdk/self").data.url("icons/icon_32.png")
			}
		});
	}else{
		button.state("window",{
			"icon" : {
				"16":require("sdk/self").data.url("icons/icon_16_off.png"),
				"32":require("sdk/self").data.url("icons/icon_32_off.png")
			}
		});
	}
}

if (typeof jjDict === "undefined") {
	var { ToggleButton } = require("sdk/ui/button/toggle");
    var data = require("sdk/self").data;
    var request = require("sdk/request").Request;
    var prefs =  require("sdk/simple-prefs").prefs;
    var system = require("sdk/system");

	var buttonPanel = require("sdk/panel").Panel({
		contentURL: require("sdk/self").data.url("buttonPanel.html"),
		onHide:function(){
			actionButton.state('window',{checked:false});
			iconStateChange(jjDict.button,jjDict.prefs["jjdict.enable"]);
		}
	});

	var actionButton = ToggleButton({
		id: "jjDicTrans-button",
		label: "jjDicTrans",
		icon: {
			"16":data.url("icons/icon_16.png"),
			"32":data.url("icons/icon_32.png")
		},
		onChange: function(state){
			if(state.checked){
				buttonPanel.show({
					position:actionButton,
					width:220,
					height:450
				});
				buttonPanel.port.emit("init_load",jjDict.prefs);
				buttonPanel.port.on("update_prefs",function(type,value){
					if(type=="enable"){
						jjDict.prefs["jjdict.enable"] = value;
						iconStateChange(jjDict.button,jjDict.prefs["jjdict.enable"]);
					}else if(type=="keys"){
						jjDict.prefs["jjdict.keys.shift"] = value["shift"];
						jjDict.prefs["jjdict.keys.ctrl"] = value["ctrl"];
						jjDict.prefs["jjdict.keys.alt"] = value["alt"];
						jjDict.prefs["jjdict.keys.other_key"] = value["other_key"];
						jjDict.prefs["jjdict.keys.dbclick"] = value["dbclick"];
					}else if(type=="backend.dict"){
						jjDict.prefs["jjdict.backend.dict"] = value;
					}else if(type=="backend.trans"){
						jjDict.prefs["jjdict.backend.trans"] = value;
					}else if(type=="mode"){
						jjDict.prefs["jjdict.mode"] = value;
					}else if(type=="popup_pos"){
						jjDict.prefs["jjdict.popup_pos"] = value;
					}
				});
				buttonPanel.port.on("show_main_panel",function(type,value){
						if(type == "dict"){
						}else if(type = "trans"){
						}
				});
				buttonPanel.port.on("open_panel",function(type, keyword){
					if(type == "dict"){
						var backend = backendSet[p['jjdict.lang.from']].dict[p['jjdict.backend.dict']];
					}else{
						var backend = backendSet[p['jjdict.lang.from']].trans[p['jjdict.backend.trans']];
					}
			      var t_url = backend.url(keyword);
					var pos = backend.size;
					var url = t_url[0];
					var text = t_url[1];

			  	var script = "";
			  	if(text && text.oid && text.oid.length > 0){
			          script = "self.port.on('showtext',function(text){";
			          script +=" 	  document.getElementById('"+text.oid+"').value = text; ";
			          if(text.btn && text.btn.length > 0){
			          	script += "    document.getElementById('"+text.btn+"').click()";
			          }
			          script +="});";
			  	}

					if(jjDict.panel) jjDict.panel.destroy();
					jjDict.panel = require("sdk/panel").Panel({
						contentURL:jjDict.data.url(url),
						contentStyle:"body { background-color:white; border: 1px solid #EEEEEE; }",
						contentScriptFile: jjDict.data.url('mainPanel.js'),
						contentScript:script
					});
					jjDict.panel.port.emit("init_platform",jjDict.system.platform);
					jjDict.panel.show({
						width:pos.w,
						height:pos.h,
						position:jjDict.button
					});
		    	if(text && text.oid && text.oid.length > 0){
			    	jjDict.panel.port.emit("showtext",text.text);
					}
				});
			}
		}
	});

 var jjDict = {
    data : data,
    request : request,
    prefs : prefs,
    panel: false,
    system: system,
    detachWorker : function (worker, workerArray) {
      var index = workerArray.indexOf(worker);
      if(index !== -1) {
        workerArray.splice(index, 1);
      }
    },
    button:actionButton,
    buttonPanel:buttonPanel,
    work:false
  };
};

var p = jjDict.prefs;
// 초기 값 셋팅
{
	if(p["jjdict.enable"] == undefined || p["jjdict.enable"] == ""){p["jjdict.enable"] = true; }
	if(p["jjdict.mode"] == undefined || p["jjdict.mode"] == ""){p["jjdict.mode"] = 0; }
	if(p["jjdict.popup_pos"] == undefined || p["jjdict.popup_pos"] == ""){p["jjdict.popup_pos"] = 0; }
	if(p["jjdict.keys.dbclick"] == undefined || p["jjdict.keys.dbclick"] == ""){p["jjdict.keys.dbclick"] = false; }
	if(p["jjdict.keys.shift"] == undefined || p["jjdict.keys.shift"] == ""){p["jjdict.keys.shift"] = false; }
	if(p["jjdict.keys.ctrl"] == undefined || p["jjdict.keys.ctrl"] == ""){p["jjdict.keys.ctrl"] = false; }
	if(p["jjdict.keys.alt"] == undefined || p["jjdict.keys.alt"] == ""){p["jjdict.keys.alt"] = false; }
	if(p["jjdict.keys.other_key"] == undefined || p["jjdict.keys.other_key"] == ""){p["jjdict.keys.other_key"] = ""; }
	if(p["jjdict.lang.from"] == undefined || p["jjdict.lang.from"] == ""){p["jjdict.lang.from"] = 'en'; }
	if(p["jjdict.lang.to"] == undefined || p["jjdict.lang.to"] == ""){p["jjdict.lang.to"] = 'ko'; }
	if(p["jjdict.backend.dict"] == undefined || p["jjdict.backend.dict"] == ""){p["jjdict.backend.dict"] = "daum"; }
	if(p["jjdict.backend.trans"] == undefined || p["jjdict.backend.trans"] == ""){p["jjdict.backend.trans"] = "google"; }
	iconStateChange(jjDict.button,p["jjdict.enable"]);

	if( p["jjdict.keys.shift"] == false &&
		p["jjdict.keys.ctrl"] == false &&
		p["jjdict.keys.alt"] == false &&
		p["jjdict.keys.dbclick"] == false &&
		p["jjdict.keys.other_key"] == "")
	{
		p["jjdict.keys.shift"] = true; // Default Activation Key
	}
}

var cm = require("sdk/context-menu");
cm.Item({
  label: "Search jjDicTrans",
  context: cm.SelectionContext(),
  contentScript: 'self.on("context", function () {' +
                 '  var text = window.getSelection().toString();' +
                 '  if (text.length > 17)' +
                 '    text = text.substr(0, 17) + "...";' +
                 '  return "Search jjDicTrans for : \\"" + text + "\\""; });' +
				'self.on("click", function (node,data) {' +
                 '  var text = window.getSelection().toString();' +
                 '  self.postMessage(text);' +
                 '});',
	onMessage: function(text){
		for(var i in workers){
			var work = workers[i];
			work.port.emit('jjdict.context_click',text);
		}
	}
});

require("sdk/page-mod").PageMod({
	include: ['*'],
	contentScriptWhen: 'ready',
	contentScriptFile: jjDict.data.url('jjdict_main.js'),
	contentStyleFile: jjDict.data.url('css/style.css'),
	attachTo: ["top", "frame", "existing"],
	onAttach: function(work) {
		workers.push(work);
		work.on('detach',function(){
			detachWorker(this,workers);
		});
		work.port.emit('jjdict.conf', {
			enable:p['jjdict.enable'],
			mode:p['jjdict.mode'],
			popup_pos:p['jjdict.popup_pos'],
			shift:p['jjdict.keys.shift'],
			ctrl:p['jjdict.keys.ctrl'],
			alt:p['jjdict.keys.alt'],
			other_key:p['jjdict.keys.other_key'],
			dbclick:p['jjdict.keys.dbclick'],
			lang_from:p['jjdict.lang.from'],
			lang_to:p['jjdict.lang.to'],
			backend_dict:backendSet[p['jjdict.lang.from']].dict[p['jjdict.backend.dict']],
			backend_trans:backendSet[p['jjdict.lang.from']].trans[p['jjdict.backend.trans']]
		});

		require("sdk/simple-prefs").on("", function(){
			for(var i in workers){
				var twork = workers[i];
				twork.port.emit('jjdict.conf', {
					enable:p['jjdict.enable'],
					mode:p['jjdict.mode'],
					popup_pos:p['jjdict.popup_pos'],
					shift:p['jjdict.keys.shift'],
					ctrl:p['jjdict.keys.ctrl'],
					alt:p['jjdict.keys.alt'],
					other_key:p['jjdict.keys.other_key'],
					dbclick:p['jjdict.keys.dbclick'],
					lang_from:p['jjdict.lang.from'],
					lang_to:p['jjdict.lang.to'],
					backend_dict:backendSet[p['jjdict.lang.from']].dict[p['jjdict.backend.dict']],
					backend_trans:backendSet[p['jjdict.lang.from']].trans[p['jjdict.backend.trans']]
				});
			}
		});

		work.port.on('jjdict.request', function(text,type,pos){
			// URL Make
			if(type == "dict"){
				var backend = backendSet[p['jjdict.lang.from']].dict[p['jjdict.backend.dict']];
			}else{
				var backend = backendSet[p['jjdict.lang.from']].trans[p['jjdict.backend.trans']];
			}
			var t_url = backend.url(text);
			var sz = backend.size;
			var url = t_url[0];
			var text = t_url[1];

			if(!pos || pos == undefined || p['jjdict.popup_pos'] == 0){
				pos = jjDict.button;
			}else{
				tleft = pos.left - (sz.w/2);
				ttop = pos.top - (sz.h + 15);
				if(ttop < 0){ ttop = pos.top + 15; }
				pos.left = tleft;
				pos.top = ttop;
			}

		  	var script = "";
		  	if(text && text.oid && text.oid.length > 0){
		          script = "self.port.on('showtext',function(text){";
		          script +=" 	  document.getElementById('"+text.oid+"').value = text; ";
		          if(text.btn && text.btn.length > 0){
		          	script += "    document.getElementById('"+text.btn+"').click()";
		          }
		          script +="});";
		  	}

			if(jjDict.panel) jjDict.panel.destroy();
			jjDict.panel = require("sdk/panel").Panel({
				contentURL:jjDict.data.url(url),
				contentStyle:"body { background-color:white; border: 1px solid #EEEEEE; overflow-x:hidden; }",
				contentScriptFile: jjDict.data.url('mainPanel.js'),
				contentScript:script
			});
			jjDict.panel.port.emit("init_platform",jjDict.system.platform);
			jjDict.panel.show({
				width:sz.w,
				height:sz.h,
				position:pos
			});
	    	if(text && text.oid && text.oid.length > 0){
	    		jjDict.panel.port.emit("showtext",text.text);
			}
		});

	}
});


data/mainPanel.js
function initAudio() {
	var audio = document.createElement('AUDIO');
	audio.id = 'my_player';
	var body = document.getElementsByTagName('BODY')[0];
	body.appendChild(audio);

	return audio;
}

function removeElement(querys) {
	for (var i in querys) {
		var query = querys[i];
		var obj = document.querySelector(query);
		if (obj) { obj.style.display = 'none'; }
	}
}

function removeBackground(querys) {
	for (var i in querys) {
		var query = querys[i];
		var obj = document.querySelector(query);
		if (obj) { obj.style.background = 'none'; }
	}
}

function daumDictModify(audio) {
	//removeElement(['header']);
	//removeBackground(['#daumWrap']);
	//var btn = document.querySelector('.btn_before');
	//btn.addEventListener('onclick', function(){
	//	console.log("alskdjflaksdjflaskdjf");
	//}, false);
	return;
}

function naverDictModify(audio) {
	//removeElement(['#header','#top_search']);
	//removeElement(['#header']);
}

function googleTransModify(audio) {
	//removeElement(['#gb','#gba']);
}

self.port.on('init_platform',function(platform) {
	// Mac = darwin
	// Linux = linux
	// Windows = winnt
	var audio = initAudio();

	document.documentElement.style.overflowX = 'hidden';

	// 1. is daum?
	var isDaum = (document.location.host.indexOf('daum') != -1) ? true : false;
	// 2. is naver?
	var isNaver = (document.location.host.indexOf('naver') != -1) ? true : false;
	// 3. is google?
	var isGoogle = (document.location.host.indexOf('google') != -1) ? true : false;

	/*
	if (platform == 'winnt') {
		if (isDaum) { daumDictModify(audio); }
		if (isNaver) { naverDictModify(audio,true); }
	}else if (platform == 'linux') {
		//if (is_daum) daum_dict_modify(audio);
		//if (is_naver) naver_dict_modify(audio,true);
	}else if (platform == 'darwin') {
		if (isNaver) { naverDictModify(audio,false); }
	}
	*/
	if (isDaum) { daumDictModify(audio); }
	if (isNaver) { naverDictModify(audio); }
	if (isGoogle) { googleTransModify(audio); }
});


data/buttonPanel.js
var p_other_key = "";

function enable_clicked(){
	var enable = document.getElementById("enable");
	var enable_value = false;
	if(enable.checked == true){
		enable_value = true;
	}
	addon.port.emit('update_prefs',"enable",enable_value);
}

function all_key_unchecked(shift,ctrl,alt,other_key,dbclick){
	//console.log("shift="+shift.checked+", ctrl="+ctrl.checked+", alt="+alt.checked+", other_key="+other_key.value );
	if(shift.checked == false &&
		ctrl.checked == false &&
		alt.checked == false &&
		dbclick.checked == false &&
		other_key.value == "")
	{
		return true;
	}
	return false;
}

function get_code_from_string(str){
	if(str.length < 1) return -1;
	if(str == "Backspace"){ return 8; }
	if(str == "Tab"){ return 9; }
	if(str == "Enter"){ return 13; }
	if(str == "Shift"){ return 16; }
	if(str == "Ctrl"){ return 17; }
	if(str == "Alt"){ return 18; }
	if(str == "Pause/break"){ return 19; }
	if(str == "Caps lock"){ return 20; }
	if(str == "Escape"){ return 27; }
	if(str == "Page up"){ return 33; }
	if(str == "Page down"){ return 34; }
	if(str == "End"){ return 35; }
	if(str == "Home"){ return 36; }
	if(str == "Left arrow"){ return 37; }
	if(str == "Up arrow"){ return 38; }
	if(str == "Right arrow"){ return 39; }
	if(str == "Down arrow"){ return 40; }
	if(str == "Insert"){ return 45; }
	if(str == "Delete"){ return 46; }
	if(str == "Left window"){ return 91; }
	if(str == "Right window"){ return 92; }
	if(str == "Select key"){ return 93; }
	if(str == "Numpad 0"){ return 96; }
	if(str == "Numpad 1"){ return 97; }
	if(str == "Numpad 2"){ return 98; }
	if(str == "Numpad 3"){ return 99; }
	if(str == "Numpad 4"){ return 100; }
	if(str == "Numpad 5"){ return 101; }
	if(str == "Numpad 6"){ return 102; }
	if(str == "Numpad 7"){ return 103; }
	if(str == "Numpad 8"){ return 104; }
	if(str == "Numpad 9"){ return 105; }
	if(str == "Multiply"){ return 106; }
	if(str == "Add"){ return 107; }
	if(str == "Subtract"){ return 109; }
	if(str == "Decimal point"){ return 110; }
	if(str == "Divide"){ return 111; }
	if(str == "F1"){ return 112; }
	if(str == "F2"){ return 113; }
	if(str == "F3"){ return 114; }
	if(str == "F4"){ return 115; }
	if(str == "F5"){ return 116; }
	if(str == "F6"){ return 117; }
	if(str == "F7"){ return 118; }
	if(str == "F8"){ return 119; }
	if(str == "F9"){ return 120; }
	if(str == "F10"){ return 121; }
	if(str == "F11"){ return 122; }
	if(str == "F12"){ return 123; }
	if(str == "Num lock"){ return 144; }
	if(str == "Scroll lock"){ return 145; }
	if(str == ";"){ return 186; }
	if(str == "="){ return 187; }
	if(str == ","){ return 188; }
	if(str == "-"){ return 189; }
	if(str == "."){ return 190; }
	if(str == "/"){ return 191; }
	if(str == "`"){ return 192; }
	if(str == "["){ return 219; }
	if(str == "\\"){ return 220; }
	if(str == "]"){ return 221; }
	if(str == "'"){ return 222; }

	return str.charCodeAt(0);
}

function get_string_from_code(keyCode){
	if(isNaN(keyCode) || keyCode=="" || keyCode == -1){
		return "";
	}
	var ret = String.fromCharCode(keyCode);
	if (keyCode == 8) ret = "Backspace";
	if (keyCode == 9) ret = "Tab";
	if (keyCode == 13) ret = "Enter";
	if (keyCode == 16) ret = "Shift";
	if (keyCode == 17) ret = "Ctrl";
	if (keyCode == 18) ret = "Alt";
	if (keyCode == 19) ret = "Pause/break";
	if (keyCode == 20) ret = "Caps lock";
	if (keyCode == 27) ret = "Escape";
	if (keyCode == 33) ret = "Page up";
	if (keyCode == 34) ret = "Page down";
	if (keyCode == 35) ret = "End";
	if (keyCode == 36) ret = "Home";
	if (keyCode == 37) ret = "Left arrow";
	if (keyCode == 38) ret = "Up arrow";
	if (keyCode == 39) ret = "Right arrow";
	if (keyCode == 40) ret = "Down arrow";
	if (keyCode == 45) ret = "Insert";
	if (keyCode == 46) ret = "Delete";
	if (keyCode == 91) ret = "Left window";
	if (keyCode == 92) ret = "Right window";
	if (keyCode == 93) ret = "Select key";
	if (keyCode == 96) ret = "Numpad 0";
	if (keyCode == 97) ret = "Numpad 1";
	if (keyCode == 98) ret = "Numpad 2";
	if (keyCode == 99) ret = "Numpad 3";
	if (keyCode == 100) ret = "Numpad 4";
	if (keyCode == 101) ret = "Numpad 5";
	if (keyCode == 102) ret = "Numpad 6";
	if (keyCode == 103) ret = "Numpad 7";
	if (keyCode == 104) ret = "Numpad 8";
	if (keyCode == 105) ret = "Numpad 9";
	if (keyCode == 106) ret = "Multiply";
	if (keyCode == 107) ret = "Add";
	if (keyCode == 109) ret = "Subtract";
	if (keyCode == 110) ret = "Decimal point";
	if (keyCode == 111) ret = "Divide";
	if (keyCode == 112) ret = "F1";
	if (keyCode == 113) ret = "F2";
	if (keyCode == 114) ret = "F3";
	if (keyCode == 115) ret = "F4";
	if (keyCode == 116) ret = "F5";
	if (keyCode == 117) ret = "F6";
	if (keyCode == 118) ret = "F7";
	if (keyCode == 119) ret = "F8";
	if (keyCode == 120) ret = "F9";
	if (keyCode == 121) ret = "F10";
	if (keyCode == 122) ret = "F11";
	if (keyCode == 123) ret = "F12";
	if (keyCode == 144) ret = "Num lock";
	if (keyCode == 145) ret = "Scroll lock";
	if (keyCode == 186) ret = ";";
	if (keyCode == 187) ret = "=";
	if (keyCode == 188) ret = ",";
	if (keyCode == 189) ret = "-";
	if (keyCode == 190) ret = ".";
	if (keyCode == 191) ret = "/";
	if (keyCode == 192) ret = "`";
	if (keyCode == 219) ret = "[";
	if (keyCode == 220) ret = "\\";
	if (keyCode == 221) ret = "]";
	if (keyCode == 222) ret = "'";
	return ret;
}

function link_clicked(type,data){
	addon.port.emit("show_main_panel",type,data);
}

function activation_key_clicked(key){
	var shift = document.getElementById("shift");
	var ctrl = document.getElementById("ctrl");
	var alt = document.getElementById("alt");
	var other_key = document.getElementById("other_key");
	var dbclick = document.getElementById("dbclick");

	if(all_key_unchecked(shift,ctrl,alt,other_key,dbclick)){
		//console.log("all_key_unchecked1="+key);
		if(key == "shift"){ shift.checked = true; }
		if(key == "ctrl"){ ctrl.checked = true; }
		if(key == "alt"){ alt.checked = true; }
		if(key == "dbclick"){ dbclick.checked = true; }
	}

	addon.port.emit('update_prefs',"keys",{shift:shift.checked,ctrl:ctrl.checked,alt:alt.checked,other_key:get_code_from_string(other_key.value),dbclick:dbclick.checked});
}

function activation_other_key(event){
	if(event.defaultPrevented){ return; }

	var shift = document.getElementById("shift");
	var ctrl = document.getElementById("ctrl");
	var alt = document.getElementById("alt");
	var other_key = document.getElementById("other_key");
	var dbclick = document.getElementById("dbclick");

	if(event.keyCode == 8 || event.keyCode == 16 || event.keyCode == 17 || event.keyCode == 18  ) {  //Backspace, Shift, Ctrl, Alt
		other_key.value = "";
	}else{
		other_key.value = get_string_from_code(event.keyCode);
	}

	if(all_key_unchecked(shift,ctrl,alt,other_key,dbclick)){
		//console.log("all_key_unchecked="+p_other_key);
		other_key.value = p_other_key;
	}

	p_other_key = other_key.value;

	addon.port.emit('update_prefs',"keys",{shift:shift.checked,ctrl:ctrl.checked,alt:alt.checked,other_key:get_code_from_string(other_key.value),dbclick:dbclick.checked});
}

function other_key_focused(e){
	var other_key = document.getElementById("other_key");
	other_key.value = "";
}

function other_key_blured(e){
	var other_key = document.getElementById("other_key");
	other_key = p_other_key;
}


function backend_dict_changed(){
	var backend_dict = document.getElementById("backend_dict");
	for(var i=0; i < backend_dict.options.length; i++){
		var opt = backend_dict.options[i];
		if(opt.selected == true){
			addon.port.emit('update_prefs',"backend.dict",opt.value);
		}
	}
}
function backend_trans_changed(){
	var backend_trans = document.getElementById("backend_trans");
	for(var i=0; i < backend_trans.options.length; i++){
		var opt = backend_trans.options[i];
		if(opt.selected == true){
			addon.port.emit('update_prefs',"backend.trans",opt.value);
		}
	}
}

function mode_changed(){
	var mode = document.getElementById("mode");
	for(var i=0; i < mode.options.length; i++){
		var opt = mode.options[i];
		if(opt.selected == true){
			addon.port.emit('update_prefs',"mode",opt.value);
		}
	}
}

function popup_pos_changed(){
	var popup_pos = document.getElementById("popup_pos");
	for(var i=0; i < popup_pos.options.length; i++){
		var opt = popup_pos.options[i];
		if(opt.selected == true){
			addon.port.emit('update_prefs',"popup_pos",opt.value);
		}
	}
}

function open_panel(type){
		var keyword = document.getElementById("keyword");
		var keywordText = keyword.value;

		if(keywordText.length > 0){
			addon.port.emit('open_panel',type, keywordText);
		} else {
			addon.port.emit('open_panel',type, "");
		}
}
addon.port.on("init_load",function(prefs){
	var enable = document.getElementById("enable");
	if(prefs["jjdict.enable"] == true){
		enable.checked = true;
	}else{
		enable.checked = false;
	}

	var shift = document.getElementById("shift");
	var ctrl = document.getElementById("ctrl");
	var alt = document.getElementById("alt");
	var other_key = document.getElementById("other_key");
	var dbclick = document.getElementById("dbclick");

	if(prefs["jjdict.keys.shift"] == true){ shift.checked = true; }
	if(prefs["jjdict.keys.ctrl"] == true){ ctrl.checked = true; }
	if(prefs["jjdict.keys.alt"] == true){ alt.checked = true; }
	if(prefs["jjdict.dbclick"] == true){ dbclick.checked = true; }

	p_other_key = get_string_from_code(prefs["jjdict.keys.other_key"]);
	other_key.value = p_other_key;

	var mode = document.getElementById("mode");
	for(var i=0; i < mode.options.length; i++){
		var opt = mode.options[i];
		if(opt.value == prefs["jjdict.mode"]){
			opt.selected = true;
		}else{
			opt.selected = false;
		}
	}

	var backend_dict = document.getElementById("backend_dict");
	for(var i=0; i < backend_dict.options.length; i++){
		var opt = backend_dict.options[i];
		if(opt.value == prefs["jjdict.backend.dict"]){
			opt.selected = true;
		}else{
			opt.selected = false;
		}
	}

	var backend_trans = document.getElementById("backend_trans");
	for(var i=0; i < backend_trans.options.length; i++){
		var opt = backend_trans.options[i];
		if(opt.value == prefs["jjdict.backend.trans"]){
			opt.selected = true;
		}else{
			opt.selected = false;
		}
	}

	var popup_pos = document.getElementById("popup_pos");
	for(var i=0; i < popup_pos.options.length; i++){
		var opt = popup_pos.options[i];
		if(opt.value == prefs["jjdict.popup_pos"]){
			opt.selected = true;
		}else{
			opt.selected = false;
		}
	}

	// Focus keyword input
	var keyword = document.getElementById("keyword");
	keyword.focus();
	keyword.addEventListener("keyup", function(event) {
		event.preventDefault();
		if (event.keyCode == 13) {
			open_panel('dict');
		}
	});
});
